File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.java
class Solution {
    public int firstStableIndex(int[] nums, int k) {
        int n=nums.length;
        int [] sufixmin=new int[n];
        sufixmin[n-1]=nums[n-1];
        for(int i=n-2;i>=0;i--)
        {
            sufixmin[i]=Math.min(sufixmin[i+1],nums[i]);
        }
        int maxprefix=nums[0];
        for(int i=0;i<n;i++)
        {
            maxprefix=Math.max(maxprefix,nums[i]);
           int score=maxprefix - sufixmin[i];
           if(score<=k)
            return i;
        }
        return -1;
    }
}

File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.cpp
class Solution {
public:
    int firstStableIndex(vector<int>& nums, int k) {
    int n=nums.size();
    vector<int>minsufix(n);
     minsufix[n-1]=nums[n-1];
    for(int i=n-2;i>=0;i--)
    {
        minsufix[i]=min(nums[i],minsufix[i+1]);
    }

    int maxprefix=nums[0];
    for(int i=0;i<n;i++)
    {
        maxprefix=max(nums[i],maxprefix);
        if(maxprefix-minsufix[i]<=k)
            return i;
    }
    return -1;
    }
};